doc/specs/stdlib_hash_procedures.md

```diff
@@ -0,0 +1,1748 @@
+the *nmhash32* and *nmhash32x* of James Z. M. Gao;
+and the *waterhash*  of Tommy Ettinger.
+The detailed implementation of each algorithm is handled in a separate
+submodule: `stdlib_32_bit_fnv_hashes`,
+`stdlib_32_bit_nmhashes`, and `stdlib_32_bit_water_hashes`,
+respectively. The `nmhash32`, `nmhash32x`, and `waterhash` algorithms
+require seeds. The submodules provide separate seed generators
+for each algorithm.
+The module itself
+implements two scalar hash functions, `FIBONACCI_HASH` and
+`UNIVERSAL_MULT_HASH`. 
+It also implements the subroutine, `ODD_RANDOM_INTEGER`, for
+generating seeds for `UNIVERSAL_MULT_HASH`.
+All assume a two's complement sign bit, and no out of
+range checks.
+
+The `stdlib_64_bit_hash_functions` module also provides
+implementations of four hash code algorithms:
+the *FNV_1* and *FNV_1A* variants of Glenn Fowler, 
+Landon Curt Noll, and Kiem-Phong Vo;
+the *pengyhash* of Alberto Fajardo;
+and the *SpookyHash*  of Bob Jenkins.
+The detailed implementation of each algorithm is handled in a separate
+submodule: `stdlib_64_bit_fnv_hashes`,
+`stdlib_64_bit_pengy_hashes`, and `stdlib_64_bit_spooky_hashes`,
+respectively.
+The `pengyhash`, and `Spooky Hash` algorithms
+require seeds. The submodules provide separate seed generators
+for each algorithm.
+The module itself implements two scalar hash functions,
+`FIBONACCI_HASH` and `UNIVERSAL_MULT_HASH`.
+It also implements the subroutine, `ODD_RANDOM_INTEGER`, for
+generating seeds for `UNIVERSAL_MULT_HASH`.
+All assume a two's complement sign bit, and no out of
+range checks.
+
+The `stdlib_32_bit_fnv_hashes` and `stdlib_64_bits_fnv_hashes`
+submodules each provide implementations of the FNV-1 and FNV-1A
+algorithms in the form of two separate overloaded functions: `FNV_1`
+and `FNV_1A`.
+The FNV-1 and FNV-2 algorithms differ in their order of the
+multiplication and exclusive or operations.
+They differ from their normal implementation in that they also
+encode the structure size in the hash code.
+The 32 and 64 bit algorithms differ in their initial offsets and in
+their multiplicative constants.
+Analysis suggests that `FNV_1A` should be better at randomizing the
+input, but tests with hash tables show negligible difference.
+These algorithms have the reputation of being particularly useful for
+small byte strings, i.e., strings of less than 32 bytes.
+While they do not at all perform well on the SMHasher test suite,
+usage indicates that this has little impact on the
+performance of small hash tables, and the small size of the functions
+allows their quick loading and retainment in the instruction cache,
+giving a performance boost where the hashing is intermittent.
+(See the
+[SMHasher discussion](https://github.com/rurban/smhasher/README.md)
+and S. Richter, V. Alvarez, and J. Dittrich,
+["A Seven-Dimensional Analysis of Hashing Methods and its Implications on Query Processing"](https://bigdata.uni-saarland.de/publications/p249-richter.pdf).
```

Comment:
Do we have an internet archive link for this?

Reply:
Was thinking about [internet archive](https://archive.org/), but a doi would work as well.

Reply:
Should I replace a link to the full paper with a dog link that omits the main body of the article?

Reply:
This paper doesn't have a `doi`. I suggest you provide the full reference as well as the current link - and if the latter fails at some point at least the paper will be findable.

` S. Richter, V. Alvarez, and J. Dittrich, 2015, A Seven-Dimensional Analysis of Hashing Methods and its Implications on Query Processing, Proceedings of the VLDB Endowment, Vol. 9, No. 3`

Reply:
@gareth-nx Is this DOI wrong: `https://doi.org/10.14778/2850583.2850585`? 
The full reference is ok too.

Reply:
Ahh -- my mistake. In that case I suggest just using the doi url, because it should be long-term stable